### ISSUE TITLE
index out-of-bounds error

### ISSUE THREAD

**Issue by tubixiansheng** [iSEE-Laboratory/ReferDINO#13]
Hello, if I want to perform a multi-class task, should I change the following code: `layer_class = torch.max(layer_cls_embed(layer_hs, text_dict), dim=-1, keepdim=True)[0]` to: `layer_class = layer_cls_embed(layer_hs, text_dict) # shape: (N, num_classes)`? Otherwise, the matcher will report an index out-of-bounds error.

**Comment by tmliang** [comment]
Hi, the current version only supports single-class training, because the matcher and criterion only consider the single-class cases:
https://github.com/iSEE-Laboratory/ReferDINO/blob/64e4a4a7211947db2c6f40f072f604ab2fefc308/models/matcher.py#L172-L173
So if you need multi-class training, I think you have to modify at least the 1) model output, 2) matcher and criterion, 3) the dataloader to provide a label mapping. 

I think you might refer to the training code of [OpenGroundingDINO](https://github.com/longzw1997/Open-GroundingDino), which would be helpful for these modifications.

**Comment by tubixiansheng** [comment]
> Hi, the current version only supports single-class training, because the matcher and criterion only consider the single-class cases:
> 
> [ReferDINO/models/matcher.py](https://github.com/iSEE-Laboratory/ReferDINO/blob/64e4a4a7211947db2c6f40f072f604ab2fefc308/models/matcher.py#L172-L173)
> 
> Lines 172 to 173 in [64e4a4a](/iSEE-Laboratory/ReferDINO/commit/64e4a4a7211947db2c6f40f072f604ab2fefc308)
> 
>  # Only has one tgt, MinCost Matcher 
>  _, src_ind = torch.min(C, dim=0) 
> 
> So if you need multi-class training, I think you have to modify at least the 1) model output, 2) matcher and criterion, 3) the dataloader to provide a label mapping.
> I think you might refer to the training code of [OpenGroundingDINO](https://github.com/longzw1997/Open-GroundingDino), which would be helpful for these modifications.

I understand now, thank you for your reply.

**Comment by tmliang** [comment]
BTW, another way is to keep the single-class training pipeline, but perform multi-class inference. This will be much easier. The only part that needs to be changed is in the inference code, where we retain all objects with socres above the threshold, instead of considering only the highest one, like 

https://github.com/iSEE-Laboratory/ReferDINO/blob/64e4a4a7211947db2c6f40f072f604ab2fefc308/eval/inference_mevis.py#L242-L248

**Comment by tubixiansheng** [comment]
> BTW, another way is to keep the single-class training pipeline, but perform multi-class inference. This will be much easier. The only part that needs to be changed is in the inference code, where we retain all objects with socres above the threshold, instead of considering only the highest one, like
> 
> [ReferDINO/eval/inference_mevis.py](https://github.com/iSEE-Laboratory/ReferDINO/blob/64e4a4a7211947db2c6f40f072f604ab2fefc308/eval/inference_mevis.py#L242-L248)
> 
> Lines 242 to 248 in [64e4a4a](/iSEE-Laboratory/ReferDINO/commit/64e4a4a7211947db2c6f40f072f604ab2fefc308)
> 
>  idx = pred_scores.squeeze(-1) > args.obj_threshold  # n 
>  if args.top1 or not idx.any(): 
>      top_score, idx = pred_scores.squeeze(-1).topk(1, sorted=False) 
>   
>  pred_masks = pred_masks[:, idx, :, :].cpu()  # [t, n, h, w] 
>  pred_boxes = pred_boxes[:, idx, :].cpu()  # [t, n, 4] 
>  pred_logits = pred_logits[:, idx, :].cpu()  # [t, n, k]

Okay, I'll try it. Thank you very much.
